src/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -261,3 +261,67 @@ function chol_blocked_rev(Σ̄::AbstractMatrix, L::AbstractMatrix, nb::Integer,
     # Convert to `Matrix`s because blas functions require StridedMatrix input.
     return chol_blocked_rev!(Matrix(Σ̄), Matrix(L), nb, upper)
 end
+
+#####
+##### `qr`
+#####
+
+const QR_TYPE = Union{QR, QRPivoted, LinearAlgebra.QRCompactWY}
+
+function rrule(::typeof(qr), X::AbstractMatrix{<:Real})
+    F = qr(X)
+
+    if size(F.R, 2) != size(F.R, 1)
+        throw(ArgumentError("Pullback for QR decomposition is only supported for m × n matrices with m >= n"))
+    end
+    function qr_pullback(Ȳ::Composite)
+        # `getproperty` on `Composite`s ensures we have no thunks.
+        ∂X = qr_rev(F, Ȳ.Q, Ȳ.R)
+        return (NO_FIELDS, ∂X)
+    end
+    return F, qr_pullback
+end
+
+function rrule(::typeof(getproperty), F::T, x::Symbol) where T <: QR_TYPE
+    function getproperty_qr_pullback(Ȳ)
+        C = Composite{T}
+        ∂F = if x === :Q
+            C(Q=Ȳ,)
+        elseif x === :R
+            C(R=Ȳ,)
+        end
```

Comment:
Is it worth defining `C` ?
```suggestion
        ∂F = if x === :Q
            Composite{T}(Q=Ȳ,)
        elseif x === :R
            Composite{T}(R=Ȳ,)
        end
```

Reply:
I did it analog to the `svd` case. I think I could go even simpler with

```julia
∂F = Composite{T}(; x => Ȳ)
```

Should I port that kind of cleanup to `svd` and `cholesky` as well?